### PR TITLE
resolve: Feed the `def_kind` query immediately on `DefId` creation

### PR DESCRIPTION
**Pull request by petrochenkov**
Before this PR the def kind query required building HIR for no good reason, with this PR def kinds are instead assigned immediately when `DefId`s are created.

Some PRs previously refactored things to make all def kinds to be available early enough - https://github.com/rust-lang/rust/pull/118250, https://github.com/rust-lang/rust/pull/118272, https://github.com/rust-lang/rust/pull/118311.

### CODE REVIEW

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/118250.
@rustbot blocked

File: compiler/rustc_resolve/src/def_collector.rs [hunk 1]
```diff
@@ -162,18 +190,23 @@ impl<'a, 'b, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'b, 'tcx> {
...
-            return self.visit_macro_invoc(foreign_item.id);
-        }
+        let def_kind = match foreign_item.kind {
+            ForeignItemKind::Static(_, mt, _) => DefKind::Static(mt),
+            ForeignItemKind::Fn(_) => DefKind::Fn,
+            ForeignItemKind::TyAlias(_) => DefKind::ForeignTy,
+            ForeignItemKind::MacCall(_) => return self.visit_macro_invoc(foreign_item.id),
+        };
 
+        // FIXME: The namespace is incorrect for foreign types.
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:205]: Please fix it as a drive-by.

**Review by cjgillot — COMMENTED** [review]
r=me with a nit

File: compiler/rustc_resolve/src/def_collector.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/def_collector.rs]: Fixed in the second commit.

File: compiler/rustc_interface/src/queries.rs [hunk 2]
```diff
@@ -171,6 +172,9 @@ impl<'tcx> Queries<'tcx> {
                 )));
                 feed.crate_for_resolver(tcx.arena.alloc(Steal::new((krate, pre_configured_attrs))));
                 feed.output_filenames(Arc::new(outputs));
+
+                let feed = tcx.feed_local_def_id(CRATE_DEF_ID);
+                feed.def_kind(DefKind::Mod);
```
Comment (bvanjoi) [inline compiler/rustc_interface/src/queries.rs:177]: I have a question regarding incremental builds: when referring to queries such as the results of ﻿`tcx.def_kind(LocalDefId)` or `﻿tcx.def_kind(DefId)`, does this mean the results are retrieved directly here?

Comment (petrochenkov) [reply compiler/rustc_interface/src/queries.rs:177]: Here the result is put directly into the query cache.
So when `tcx.def_kind(def_id)` is called it will look into the query cache first, and will always find the result there, and will never call the query implementation (which now would panic because `providers.def_kind` is not set).
